contracts/Zauction.sol

```diff
@@ -121,6 +132,77 @@ contract ZAuction is Initializable, OwnableUpgradeable {
     );
   }
 
+  function setBuyPrice(uint256 amount, uint256 tokenId) external {
+    address owner = registrar.ownerOf(tokenId);
+    require(msg.sender == owner, "zAuction: only owner can set price");
+    buyPrice[tokenId] = amount;
+  }
+
+  /// recovers buyer's signature based on seller's proposed data and, if bid data hash matches the message hash, transfers nft and payment
+  /// @param signature type encoded message signed by the buyer
+  /// @param auctionId unique per address auction identifier chosen by seller
+  /// @param amount token amount of sale
+  /// @param tokenId token id we are transferring
+  /// @param startBlock block number at which acceptBid starts working
+  /// @param expireBlock block number at which acceptBid stops working
+  function buyNow(
+    bytes memory signature,
+    uint256 auctionId,
+    uint256 amount,
+    uint256 tokenId,
+    uint256 startBlock,
+    uint256 expireBlock
+  ) external {
+    require(startBlock <= block.number, "zAuction: auction hasn't started");
+    require(expireBlock > block.number, "zAuction: auction expired");
+    require(amount == buyPrice[tokenId], "zAuction: wrong sale price");
+    address seller = registrar.ownerOf(tokenId);
+    require(msg.sender != seller, "zAuction: cannot sell to self");
+
+    require(
+      !consumed[msg.sender][auctionId],
```

Comment:
What does this protect against? `auctionId` is a parameter to this function, which the sender defines. This check doesn't allow the same auction id to be used by the same sender twice. What is `auctionid` in this context and why is it important? Do we even need this at all?